feat: query support for geo within and intersects + dynamic GraphQL operator types

AlessioGr opened the pull request:
## Description

Old PR: #3001

- [X] I have read and understand the [CONTRIBUTING.md](../CONTRIBUTING.md) document in this repository.

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Chore (non-breaking change which does not add functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Change to the [templates](../templates/) directory (does not affect core functionality)
- [ ] Change to the [examples](../examples/) directory (does not affect core functionality)
- [ ] This change requires a documentation update

## Checklist:

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Existing test suite passes locally with my changes
- [ ] I have made corresponding changes to the documentation

DanRibbens commented:
@AlessioGr what makes this a `draft` and what is missing still to be merged?

AlessioGr commented:
> @AlessioGr what makes this a `draft` and what is missing still to be merged?

@DanRibbens I'm currently working on adding tests for the within and intersects operators. After that, I think this can be merged!

DanRibbens commented:
In the filter options of the list view I was expecting to see `intersects` and `within` as operator options:
![image](https://github.com/payloadcms/payload/assets/6434612/a9bc8a4c-aea5-41f0-bf57-885d2c2596bd)

This should be added before this PR is merged.

DanRibbens reviewed (changes requested):
One change left to make: https://github.com/payloadcms/payload/pull/3183#issuecomment-1679605686

AlessioGr commented:
> In the filter options of the list view I was expecting to see `intersects` and `within` as operator options: ![image](https://user-images.githubusercontent.com/6434612/260840854-a9bc8a4c-aea5-41f0-bf57-885d2c2596bd.png)
> 
> This should be added before this PR is merged.

@DanRibbens The issue here is that even the `near` field doesn't actually work in the admin panel filter options anyways - so I didn't even bother for `within` and `intersects`.

Given that the issue has existed before this PR, I think it's out of scope. What do you think?

AlessioGr commented:
@DanRibbens I did some more digging. This would be a simple rest API call:

```json
{
  "$and": [
    {
      "point": {
        "$near": {
          "$geometry": {
            "type": "Point",
            "coordinates": [
              1,
              -7
            ]
          },
          "$maxDistance": 1000
        }
      }
    }
  ]
}
```

and it works perfectly. However, the `WhereBuilder` in the admin panel requires queries to be in this special `and > or > and` format, which makes the query look like this:


```json
{
  "$and": [
    {
      "$or": [
        {
          "$and": [
            {
              "point": {
                "$near": {
                  "$geometry": {
                    "type": "Point",
                    "coordinates": [1, -7]
                  },
                  "$maxDistance": 1000
                }
              }
            }
          ]
        }
      ]
    }
  ]
}
```

This causes the following error:

`$geoNear, $near, and $nearSphere are not allowed in this context`


Not sure what the issue is - looks like it could be a limitation in MongoDB.

AlessioGr commented:
@DanRibbens I've fixed the issue with the WhereBuilder!

The issue was that the huge `and[][0][or][0][and][0]` Where queries the `WhereBuilder` component generates did NOT properly get flattened by `flattenWhereConstrains` before. Now they do, and it works beautifully.

Given that it works now, I've added back within and intersects to the `WhereBuilder`. It also filters and works perfectly:
![Screenshot 2023-08-16 at 00 36 39@2x](https://github.com/payloadcms/payload/assets/70709113/122a3bbe-67d7-4456-a90e-4248294a55fb)

The issue here is that the where query here is represented as an `[object Object]` in the text field.

This would be the alternative:

https://github.com/payloadcms/payload/assets/70709113/5b512380-5051-452e-81cf-90169fd5f232

can't say I prefer this.

I think this would be best addressed in a separate PR, as it needs some design work as well. `Where` queries as complex as this one cannot be represented in a one-line text field. Leaving it as `[object Object]` feels like the nicest option to me.

DanRibbens commented on src/graphql/schema/withOperators.ts:
```diff
@@ -9,122 +9,211 @@ import operators from './operators';
-    type: GraphQLType | ((field: FieldAffectingData, parentName: string) => GraphQLType);
-    operators: string[];
+    operators: {
+      operator: string;
```
If we named this `name` or `value` or something else we could improve readability later on where we have: `operator.operator` which is hard to understand.

DanRibbens commented on src/graphql/schema/withOperators.ts:
```diff
@@ -135,32 +224,73 @@ const defaults: DefaultsType = {
 
 const listOperators = ['in', 'not_in', 'all'];
 
+const gqlTypeCache = new Map();
+
+const getGqlTypeFromCache = (cacheKey: string): GraphQLType => {
+  return gqlTypeCache.get(cacheKey);
+};
+
+const storeGqlTypeInCache = (cacheKey: string, type: GraphQLType) => {
+  gqlTypeCache.set(cacheKey, type);
+};
```
I don't understand the need for the map with get and set functions instead of just using a plain old object.

DanRibbens reviewed (changes requested):
I have a couple code readability things I wanted to point out.

AlessioGr commented:
@DanRibbens I've adressed them!
